src/app-layer-htp-file.c

```diff
@@ -225,34 +261,89 @@ int HTPParseContentRange(bstr * rawvalue, HtpContentRange *range)
+    }
+    flags = FileFlowToFlags(s->f, STREAM_TOCLIENT);
+    if ((s->flags & HTP_FLAG_STORE_FILES_TS) ||
+            ((s->flags & HTP_FLAG_STORE_FILES_TX_TS) && txid == s->store_tx_id)) {
+        flags |= FILE_STORE;
+        flags &= ~FILE_NOSTORE;
+    } else if (!(flags & FILE_STORE) && (s->f->file_flags & FLOWFILE_NO_STORE_TC)) {
+        flags |= FILE_NOSTORE;
+    }
+
     FileContainer * files = s->files_tc;
     if (files == NULL) {
-        SCLogDebug("no files in state");
+        s->files_tc = FileContainerAlloc();
+        if (s->files_tc == NULL) {
+            SCReturnInt(-1);
+        }
+        files = s->files_tc;
+    }
+
+    if (FileOpenFileWithId(files, &s->cfg->response.sbcfg, s->file_track_id++, filename,
+                filename_len, data, data_len, flags) != 0) {
         SCReturnInt(-1);
     }
+    FileSetTx(files->tail, txid);
 
-    HtpContentRange crparsed;
-    if (HTPParseContentRange(rawvalue, &crparsed) != 0) {
-        SCLogDebug("parsing range failed");
-        SCReturnInt(-2);
+    if (FileSetRange(files, crparsed.start, crparsed.end) < 0) {
+        SCLogDebug("set range failed");
+    }
+    htp_tx_t *tx = htp_list_get(s->conn->transactions, txid);
+    if (!tx) {
+        SCReturnInt(-1);
     }
-    if (crparsed.end <= 0) {
-        SCLogDebug("negative end in range");
-        SCReturnInt(-3);
+    uint8_t *keyurl;
+    size_t keylen;
+    if (tx->request_hostname != NULL) {
+        keylen = bstr_len(tx->request_hostname) + filename_len + 1;
+        keyurl = SCMalloc(keylen);
+        if (keyurl == NULL) {
+            SCReturnInt(-1);
+        }
+        memcpy(keyurl, bstr_ptr(tx->request_hostname), bstr_len(tx->request_hostname));
+        memcpy(keyurl + bstr_len(tx->request_hostname), filename, filename_len);
+        keyurl[keylen - 1] = 0;
+    } else {
+        // do not reassemble file without host info
+        return HTPFileOpen(
+                s, filename, (uint32_t)filename_len, data, data_len, txid, STREAM_TOCLIENT);
     }
-    int retval = FileSetRange(files, crparsed.start, crparsed.end);
-    if (retval == -1) {
-        SCLogDebug("set range failed");
+    ContainerUrlRange *file_range_container = ContainerUrlRangeGet(keyurl, keylen, &s->f->lastts);
+    SCFree(keyurl);
+    if (file_range_container == NULL) {
+        // probably reached memcap
+        return HTPFileOpen(
+                s, filename, (uint32_t)filename_len, data, data_len, txid, STREAM_TOCLIENT);
     }
-    SCReturnInt(retval);
+    s->file_range = ContainerUrlRangeOpenFile(file_range_container, crparsed.start, crparsed.end,
+            crparsed.size, &s->cfg->response.sbcfg, filename, filename_len, flags);
+    if (s->file_range == NULL) {
+        // probably reached memcap
+        return HTPFileOpen(
+                s, filename, (uint32_t)filename_len, data, data_len, txid, STREAM_TOCLIENT);
+    }
+
+    if (ContainerUrlRangeAppendData(s->file_range, data, data_len) < 0) {
```

Comment:
It looks like `s->file_range` is not holding a mutex here, and `ContainerUrlRangeAppendData` doesn't seem to take it either. The pattern we use in the flow engine for example, is that lookup/creation returns a locked object. We then do our work on it and release the lock when we're done.

Reply:
> It looks like `s->file_range` is not holding a mutex here, and `ContainerUrlRangeAppendData` doesn't seem to take it either. 

Indeed

> The pattern we use in the flow engine for example, is that lookup/creation returns a locked object. We then do our work on it and release the lock when we're done.

That is what is meant to happen here
`ContainerUrlRangeAppendData` is not part of the `work` to do as it operates on `ContainerUrlRangeFile`
The work to do is to return a `ContainerUrlRangeFile` which can be used without locks as each flow will have a different one (as opposed to `ContainerUrlRange` which has the lock for lookup/creation/work)
Only one flow will have the pointer to the `FileContainer` structure while others have buffers (or just an indication to skip the content already seen)
Then `ContainerUrlRangeFileClose` uses the lock from the structure `ContainerUrlRange` to kind of synchronize together a finished range chunk...

Is it clearer ?